File: rulebook-eslint/src/rules/todo-comment.ts
import messages from '../messages.js';
import RulebookRule from './rulebook-rule.js';
import type { Rule } from 'eslint';
import type { SourceLocation } from 'estree';

/* eslint-disable @rulebook/todo-comment */
/** {@link https://hanggrian.github.io/rulebook/rules/#todo-comment|See detail} */
class TodoCommentRule extends RulebookRule {
    constructor() {
        super('todo-comment', {
            [TodoCommentRule.MSG_KEYWORD]: messages.get(TodoCommentRule.MSG_KEYWORD),
            [TodoCommentRule.MSG_SEPARATOR]: messages.get(TodoCommentRule.MSG_SEPARATOR),
        });
    }

    override create(context: Rule.RuleContext): Rule.RuleListener {
        return {
            Program() {
                context.sourceCode.getAllComments().forEach(comment => {
                    for (const match of comment.value.matchAll(TodoCommentRule.KEYWORD_REGEX)) {
                        const loc: SourceLocation | null | undefined = comment.loc;
                        if (!loc) {
                            continue;
                        }
                        context.report({
                            loc: loc,
                            messageId: TodoCommentRule.MSG_KEYWORD,
                            data: {
                                $0: match[0],
                            },
                        });
                    }
                    for (const match of comment.value.matchAll(TodoCommentRule.SEPARATOR_REGEX)) {
                        const loc: SourceLocation | null | undefined = comment.loc;
                        if (!loc) {
                            continue;
                        }
                        context.report({
                            loc: loc,
                            messageId: TodoCommentRule.MSG_SEPARATOR,
                            data: {
                                $0: match[0].slice(-1),
                            },
                        });
                    }
                });
            },
        };
    }

    private static MSG_KEYWORD: string = 'todo.comment.keyword';
    private static MSG_SEPARATOR: string = 'todo.comment.separator';

    private static KEYWORD_REGEX: RegExp =
        new RegExp('\\b(?:[Ff][Ii][Xx][Mm][Ee]|[Tt][Oo][Dd][Oo])\\b(?<!FIXME|TODO)\\b', 'g');
    private static SEPARATOR_REGEX: RegExp = new RegExp('\\b(todo|fixme)\\S', 'gi');
}

export default new TodoCommentRule();

File: rulebook-pylint/rulebook_pylint/checkers/todo_comment.py
from __future__ import annotations

from codecs import decode

from astroid.nodes import Module
from pylint.typing import TYPE_CHECKING
from regex import IGNORECASE, Match, Pattern, compile as regex

from rulebook_pylint.checkers.rulebook_checkers import RulebookFileChecker
from rulebook_pylint.messages import _Messages

if TYPE_CHECKING:
    from pylint.lint import PyLinter


# pylint: disable=todo-comment
class TodoCommentChecker(RulebookFileChecker):
    """See detail: https://hanggrian.github.io/rulebook/rules/#todo-comment"""
    _MSG_KEYWORD: str = 'todo.comment.keyword'
    _MSG_SEPARATOR: str = 'todo.comment.separator'

    _KEYWORD_REGEX: Pattern = regex(r'\b(?i:fixme|todo)(?<!FIXME|TODO)\b')
    _SEPARATOR_REGEX: Pattern = regex(r'\b(todo|fixme)\S', IGNORECASE)

    name: str = 'todo-comment'
    msgs: dict[str, tuple[str, str, str]] = _Messages.of(_MSG_KEYWORD, _MSG_SEPARATOR)

    def process_module(self, node: Module) -> None:
        with node.stream() as stream:
            for i, line in enumerate(stream):
                # obtain comment content
                line_no: int = i + 1
                parts: list[bytes] = line.split(b'#', 1)
                if len(parts) < 2:
                    continue
                comment_content: str = decode(parts[1])

                # checks for violation
                match: Match = self._KEYWORD_REGEX.search(comment_content)
                if match:
                    self.add_message(self._MSG_KEYWORD, line=line_no, args=match.group(0))
                match: Match = self._SEPARATOR_REGEX.search(comment_content)
                if match:
                    self.add_message(self._MSG_SEPARATOR, line=line_no, args=match.group(0)[-1])


def register(linter: PyLinter) -> None:
    linter.register_checker(TodoCommentChecker(linter))
